### PR TITLE
Elasticsearch

### PR DESCRIPTION
**Pull request by alexandresobolevski**
elasticsearch support for the app

### CODE REVIEW

**Comment by chriddyp** [comment]
image
- Can we move these onto one line?
- Instead of `cursor: grab` (which is used to indicate that you can drag items around), use `cursor: pointer` for the icons

**Comment by chriddyp** [comment]
What is "Show database options"? It doesn't seem to do anything when I click on it.
show hide

**Comment by chriddyp** [comment]
image
Nice! Works with https, http, locally, and remotely

**Comment by chriddyp** [comment]
image
I noticed that the descriptions don't show up unless you click on them. I don't think that behaviour is clear, especially considering that the user's first step didn't require them to click.
Can we just keep them expanded and remove their toggle-ability? I know it's verbose but that's better than people missing some critical set-up steps.

**Comment by chriddyp** [comment]
image
We can update this link to be https://plot.ly/alpha/workspace/?upload=sql now

**Comment by chriddyp** [comment]
For some reason, the app completely collapses when I try to connect to it
image

**Comment by chriddyp** [comment]
For some reason, the app completely collapses when I try to connect to it
woah

**Comment by chriddyp** [comment]
image
I think we should make this a little bit more crystal clear that the user needs to enter `https://connector.plot.ly:5000` or `http://connector.plot.ly` into this input box. Perhaps we can include it in the instructions? 
> Query data by clicking on 'import data' from plotly workspace and choose the SQL option. Enter *https://connector.plot.ly:5000* into the input box.
> Remember to keep this app running while you are making queries!

**Comment by chriddyp** [comment]
image
I think we should expand these placeholders a bit. 
- it's not totally obvious that `username` and `password` is their DB username and password and not their plotly username and password
- they're also optional, but that's not necessarily clear
- host sometimes includes `http` but sometimes it doesn't.

I think we should make these adjustments:
- Add `Elasticsearch Settings` to the top of the form.
- `username` => `username (optional)`
- `password` => `password (optional)`
- `host` => `host e.g. http://localhost`
- `port` => `port`

File: app/actions/sessions.js [hunk 1]
```diff
@@ -59,11 +59,12 @@ export function selectDatabase () {
...
         const configuration = state.sessions.getIn(
             ['list', sessionSelected, 'configuration']
-        );
+        ).toJS();
         ipcSend(
             TASKS.SELECT_DATABASE_AND_SHOW_TABLES,
             sessionSelected,
-            configuration.toJS().database
+            configuration.database,
+            configuration.database
```
Comment (chriddyp) [inline app/actions/sessions.js]: why is this here twice?

File: app/actions/sessions.js [hunk 2]
```diff
@@ -72,15 +73,21 @@ export function previewTables (tableNames) {
     return (_, getState) => {
         const state = getState();
         const sessionSelected = state.sessions.get('sessionSelected');
-        ipcSend(TASKS.PREVIEW, sessionSelected, tableNames);
+        const database = state.sessions.getIn(
+            ['list', sessionSelected, 'configuration']
+        ).toJS().database;
```
Comment (chriddyp) [inline app/actions/sessions.js]: could this just be `getIn(['list', sessionselected, 'configuration', 'database'])`?

File: app/actions/sessions.js [hunk 3]
```diff
@@ -72,15 +73,21 @@ export function previewTables (tableNames) {
...
 }
 
 export function disconnect () {
     return (_, getState) => {
         const state = getState();
         const sessionSelected = state.sessions.get('sessionSelected');
-        ipcSend(TASKS.DISCONNECT, sessionSelected);
+        const database = state.sessions.getIn(
+            ['list', sessionSelected, 'configuration']
+        ).toJS().database;
```
Comment (chriddyp) [inline app/actions/sessions.js]: similarly here

File: backend/tasks.js [hunk 4]
```diff
@@ -30,56 +30,96 @@ export const TASKS = {
...
+    // unpack sessionSelected and decide which responseManager to use
+    console.log(sessionSelected);
+    const sequelizeSessions = Object.keys(sequelizeManager.sessions);
+    const elasticSessions = Object.keys(elasticManager.sessions);
+    const isElasticSession = contains(sessionSelected, elasticSessions);
+    const isSequelizeSession = contains(sessionSelected, sequelizeSessions);
+    console.log(sequelizeSessions);
+    console.log(elasticSessions);
+    console.log(isElasticSession);
+    console.log(isSequelizeSession);
```
Comment (chriddyp) [inline backend/tasks.js]: can we remove all of these `console.log`s?

File: backend/tasks.js [hunk 5]
```diff
@@ -30,56 +30,96 @@ export const TASKS = {
...
+        console.log('using sequelize');
+        responseManager = sequelizeManager;
+    } else if (!isElasticSession && !isSequelizeSession) {
+        console.log('session does not exist');
+        // see if new elasticSearch session
+        if (message) {
+            if (message.dialect) {
+                console.log(message.dialect);
+                responseManager = message.dialect === 'elasticsearch' ? elasticManager : sequelizeManager;
+            }
```
Comment (chriddyp) [inline backend/tasks.js]: this check for a new `responseManger` seems pretty implicit. not sure if this is something to address or not, still wrapping my head around what's going on here

Comment (chriddyp) [inline backend/tasks.js]: but at a quick glance, it seems like if this task is adding a new session, then this block should be done at the top of this code block (or explicitly created by some message task) and then you could just use your `isElasticSession` and `isSequelizeSession` variable.

that's just a gut feeling, not sure if that's actually possible or not!

File: backend/tasks.js [hunk 6]
```diff
@@ -30,56 +30,96 @@ export const TASKS = {
...
+                {message: AUTHENTICATION(error)},
+                {name: 'ConnectionError'}),
+            responseSender
+        );
+    };
+
+	log(`Sending task ${task} to responseManager`, 1);
+	log(`Sending session ${sessionSelected} to responseManager`, 1);
+	log(`Sending database ${database} to responseManager`, 1);
+    log(`Sending message ${message} to responseManager`, 1);
```
Comment (chriddyp) [inline backend/tasks.js]: weird indent here

File: backend/elasticManager.js [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+    previewTables(tables, responseSender) {
+        const index = this.sessions[this.sessionSelected].config.database;
+        const types = tables;
+
+        const promises = types.map(type => {
+            return this.sessions[this.sessionSelected].search({
+                index,
+                type,
+                size: 5,
```
Comment (chriddyp) [inline backend/elasticManager.js]: nice

File: backend/elasticManager.js [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                error: null,
+                tables: null
+            });
+        });
+    }
+
+
+    showTables(responseSender) {
+        const index = this.sessions[this.sessionSelected].config.database;
+        console.log(`index : ${index}`);
```
Comment (chriddyp) [inline backend/elasticManager.js]: can we remove all of these `console.log`s?

File: backend/elasticManager.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                ),
+                responseSender
+			);
+		} else {
+            return new Promise((resolve, reject) => {
+                this.sessions[this.sessionSelected].ping({
+                    // ping usually has a 3000ms timeout
+                    requestTimeout: Infinity,
+                    // undocumented params are appended to the query string
+                    hello: 'hello elasticsearch!'
```
Comment (chriddyp) [inline backend/elasticManager.js]: is there supposed to be a username and password here? if we're not supporting username and password, can we remove it from the UI?

File: backend/elasticManager.js [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+                    {message: APP_NOT_CONNECTED},
+                    {name: 'ConnectionError'}
+                ),
+                responseSender
+			);
+		} else {
+            return new Promise((resolve, reject) => {
+                this.sessions[this.sessionSelected].ping({
+                    // ping usually has a 3000ms timeout
+                    requestTimeout: Infinity,
```
Comment (chriddyp) [inline backend/elasticManager.js]: Hm.. we probably don't want to just hang forever though. What about like 10 or 30 seconds?

File: backend/parse.js [hunk 11]
```diff
@@ -122,3 +121,55 @@ export default function (data) {
     // TODO: keys geojson and raw are currently optional
     return {columnnames, ncols, nrows, rows, geojson, raw: data};
 }
+
+
+export function parseES(data) {
```
Comment (chriddyp) [inline backend/parse.js]: it might be nice for this to be a little bit more verbose, like `parseElasticsearch`

**Comment by chriddyp** [comment]
looking good! a handful a minor things :)

**Comment by alexandresobolevski** [comment]
- [x] https://github.com/plotly/plotly-database-connector/pull/104#issuecomment-253102350
except `extend the line across the top of the page`
- [x] https://github.com/plotly/plotly-database-connector/pull/104#issuecomment-253103419
- [] https://github.com/plotly/plotly-database-connector/pull/104#issuecomment-253104202
- [] https://github.com/plotly/plotly-database-connector/pull/104#issuecomment-253104400
- [] https://github.com/plotly/plotly-database-connector/pull/104#issuecomment-253105135
- [] https://github.com/plotly/plotly-database-connector/pull/104#issuecomment-253106616

File: backend/elasticManager.js (hunk 9 above)
Comment (alexandresobolevski) [inline backend/elasticManager.js]: no its all in the `client` object above
